File: 백준/Gold/1091. 카드 섞기/카드 섞기.py



N = int(input())
P = list(map(int,input().split()))
S = list(map(int,input().split()))
pattern = []
cards = [0,1,2] * (N//3)
origin = [0,1,2] * (N//3)
cnt = 0
while cards != P:
    cnt += 1
    new_cards = [0] * N
    for i in range(N):
        new_cards[i] = cards[S[i]]
    # print(new_cards, "   ", cards)
    cards = new_cards[:]

    if cards == origin:
        print(-1)
        break
    
else:
    print(cnt)

File: 백준/Gold/1091. 카드 섞기/카드 섞기.java
import java.util.*;

public class Main {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        // 입력
        int N = sc.nextInt();
        int[] p = new int[N];
        int[] s = new int[N];
        for (int i = 0; i < N; i++) {
            p[i] = sc.nextInt();
        }
        for (int i = 0; i < N; i++) {
            s[i] = sc.nextInt();
        }
        int[] cards = new int[N];
        int[] origin = new int[N];
        for (int i = 0; i < N; i++) {
            cards[i] = i % 3;
            origin[i] = i % 3;
        }
        boolean check = true;
        int cnt = 0;
        while (!Arrays.equals(cards, p)) {
            cnt += 1;
            int[] newCards = new int[N];
            for (int i = 0; i < N; i++) {
                newCards[i] = cards[s[i]];
            }

            cards = newCards;
            if (Arrays.equals(cards, origin)) {
                System.out.println(-1);
                check = false;
                break;
            }
        }
        if (check) {
            System.out.println(cnt);
        }
    }
}
